core/src/model.cpp

```diff
@@ -18,4 +18,19 @@ std::unique_ptr<Vector> Model::apply(Vector& in) {
 
 }
 
+void Model::add_testing_data(std::vector<Vector> ins, std::vector<Vector> outs) {
```

Comment:
Should we take in begin and end iterators rather than `std::vector` (like how stdlib does)? Something like
```suggestion
void Model::add_testing_data(std::vector<Vector>::iterator input_begin, std::vector<Vector>::iterator input_end, std::vector<Vector>::iterator output_begin, std::vector<Vector>::iterator output_end) {
```

Reply:
How about we keep it as is but also add an overload for this usage?

Reply:
Sounds good!